compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/hotspot/aarch64/shenandoah/AArch64HotSpotShenandoahCompareAndSwapOp.java

```diff
@@ -0,0 +1,218 @@
+            masm.branchConditionally(AArch64Assembler.ConditionFlag.NE, done);
+            // Set lowest two bits, which will result in actual clearing the bits
+            // after the following inversion.
+            masm.orr(64, tmp1, tmp1, config.markWordLockMaskInPlace);
+            // ... and invert all bits back to get the forwarding pointer into tmp1.
+            masm.eon(64, tmp1, tmp1, zr);
+
+            // Encode tmp1 to compare against expected.
+            compress(masm, tmp1);
+
+            // Does forwarded value of fetched from-space pointer match original
+            // value of expected?
+            masm.cmp(size, tmp1, expected);
+
+            // If not, then the failure was legitimate and we're done.
+            // Branching to done with NE condition denotes failure.
+            masm.branchConditionally(AArch64Assembler.ConditionFlag.NE, done);
+
+            // Fall through to step 3. No need for step3 label.
+
+            // Step 3. We've confirmed that the value originally held in memory
+            // (now held in result) pointed to from-space version of original
+            // expected value. Try the CAS again with the from-space expected
+            // value. If it now succeeds, we're good.
+            //
+            // Note: result holds encoded from-space pointer that matches to-space
+            // object residing at expected. result is the new "expected".
+            masm.mov(size, tmp1, result);
+            emitCompareAndSwap(masm, accessKind, address, result, tmp1, newVal, memoryOrder, true);
+            // EQ flag set iff success. result holds value fetched.
+
+            // If fetched value did not equal the new expected, this could
+            // still be a false negative because some other thread may have
+            // newly overwritten the memory value with its to-space equivalent.
+            masm.branchConditionally(AArch64Assembler.ConditionFlag.EQ, done);
+
+            // In the rare case that four steps are required to perform the
+            // requested operation, the fourth step is the same as the first.
+
+            // Step 4. CAS has failed because the value most recently fetched
+            // from addr is no longer the from-space pointer held in result. If a
+            // different thread replaced the in-memory value with its equivalent
+            // to-space pointer, then CAS may still be able to succeed. The
+            // value held in the expected register has not changed.
+            //
+            // It is extremely rare we reach this point.
+
+            emitCompareAndSwap(masm, accessKind, address, result, expected, newVal, memoryOrder, setConditionFlags);
+            // EQ flag set iff success. result holds value fetched.
+
+            masm.jmp(done);
+
+            if (setConditionFlags) {
+                masm.bind(resultNullFailure);
+                // Clear zero flag to indicate failure.
+                masm.subs(32, zr, zr, 1);
```

Comment:
This doesn't appear to be a valid instruction.  GCBarrierEmissionTest and UnsafeSubstitutionsTest fail because of this on aarch64.

Reply:
Ugh. I pushed a fix - please re-test. Thank you!